### PR TITLE
[IOTDB-2083] ID table implementation

### PR DESCRIPTION
**Pull request by SilverNarcissus**
### ID table
For mass timeseries, we need a new module to improve the performance and save memory space. This PR create a new module called id table which has two main optimization:
(1) Data ingestion schema check can be done through 2 steps of hash without mmanager tree-like structure
(2) Compress device path to id (can be configured to plain or sha256) and reuse it in both id table and memetable

### JIRA
https://issues.apache.org/jira/browse/IOTDB-2083

### Design doc
https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=195727793

### Performance test
https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199528283

<hr>

This PR has:
- [x] been self-reviewed.
    - [x] concurrent read
    - [x] concurrent write
    - [x] concurrent read and write 
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. 
- [x] added or updated version, __license__, or notice information
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [x] added integration tests.
- [x] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        "insert into root.other.d1(timestamp,s0) values(2, 3.14)",
+      };
+
+  private static boolean isEnableIDTable = false;
+
+  private static String originalDeviceIDTransformationMethod = null;
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    IoTDBDescriptor.getInstance().getConfig().setAutoCreateSchemaEnabled(true);
```
Comment (ericpai) [inline integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java:111]: By using new IT framework, altering config via IoTDBDescriptor is not worked in cluster test. Please add new methods to 'org.apache.iotdb.itbase.env.BaseConfig' and implements them.

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+import java.sql.Statement;
+import java.sql.Types;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.Assert.fail;
+
+@Category({LocalStandaloneTest.class, ClusterTest.class})
```
Comment (ericpai) [inline integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java:46]: If you category IT to `ClusterTest.class`. Please make sure that it's really passed in a cluster test.

File: server/src/main/java/org/apache/iotdb/db/metadata/id_table/IDTable.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.metadata.id_table;
```
Comment (ericpai) [inline server/src/main/java/org/apache/iotdb/db/metadata/id_table/IDTable.java:20]: It's not recommanded to use underscore '_' as a part of package name. I think 'idtable' is OK.

**Review by ericpai — CHANGES_REQUESTED** [review]

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java (hunk 1 above)
Comment (SilverNarcissus) [reply integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java:111]: Thanks, this is not a cluster test, i will remove ClusterTest.class category

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java (hunk 2 above)
Comment (SilverNarcissus) [reply integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java:46]: Thanks, this is not a cluster test, i will remove ClusterTest.class category

File: server/src/main/java/org/apache/iotdb/db/metadata/id_table/IDTable.java (hunk 3 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/id_table/IDTable.java:20]: Sure, fixed~

File: integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java (hunk 1 above)
Comment (ericpai) [reply integration/src/test/java/org/apache/iotdb/db/integration/IoTDBQueryWithIDTableIT.java:111]: OK, is this new feature worked in cluster, or only well done in standalone mode in current milestone?

File: server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java [hunk 4]
```diff
@@ -619,6 +618,23 @@ protected QueryDataSet processDataQuery(QueryPlan queryPlan, QueryContext contex
       } else if (queryPlan instanceof LastQueryPlan) {
         queryDataSet = queryRouter.lastQuery((LastQueryPlan) queryPlan, context);
       } else {
+        RawDataQueryPlan plan = (RawDataQueryPlan) queryPlan;
```
Comment (SilverNarcissus) [inline server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:621]: Remove these comments~

Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/qp/executor/PlanExecutor.java:621]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/DiskSchemaManager.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.apache.iotdb.db.metadata.idtable.entry.DiskSchemaEntry;
+import org.apache.iotdb.db.utils.TestOnly;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/** This class manages IO of id table's schema entry */
+public interface DiskSchemaManager {
+
```
Comment (zyk990424) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/DiskSchemaManager.java:29]: Maybe we should stick to the naming notations and rename this to IDiskShemaManager

File: cluster/src/main/java/org/apache/iotdb/cluster/query/LocalQueryExecutor.java [hunk 6]
```diff
@@ -1043,7 +1044,12 @@ public ByteBuffer last(LastQueryRequest request)
 
     List<Pair<Boolean, TimeValuePair>> timeValuePairs =
         LastQueryExecutor.calculateLastPairForSeriesLocally(
-            partialPaths, dataTypes, queryContext, expression, request.getDeviceMeasurements());
+            partialPaths,
+            dataTypes,
+            queryContext,
+            expression,
+            request.getDeviceMeasurements(),
+            IoTDBDescriptor.getInstance().getConfig().isEnableIDTable());
```
Comment (zyk990424) [inline cluster/src/main/java/org/apache/iotdb/cluster/query/LocalQueryExecutor.java:1052]: Is there any test based on this parameter? Maybe the function implement could access the config via IoTDB global  variables.

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+      logger.debug("create new file for id table: " + dir.getName());
+      boolean createRes = dataFile.createNewFile();
+      if (!createRes) {
+        throw new IOException(
+            "create new file for id table failed. Path is: " + dataFile.getPath());
+      }
+
+      loc = 0;
+    }
+  }
```
Comment (zyk990424) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:81]: Maybe last entry check while restarting is necessary here. It should be ensured that there's no file corruption.

**Review by zyk990424 — COMMENTED** [review]
Fantastic feature and I'm looking forward to using it! There're a few suggestions.

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/DiskSchemaManager.java (hunk 5 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/DiskSchemaManager.java:29]: Fixed~

File: cluster/src/main/java/org/apache/iotdb/cluster/query/LocalQueryExecutor.java (hunk 6 above)
Comment (SilverNarcissus) [reply cluster/src/main/java/org/apache/iotdb/cluster/query/LocalQueryExecutor.java:1052]: Sure~ Fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java (hunk 7 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:81]: Fixed~

Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:81]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java [hunk 8]
```diff
@@ -129,6 +136,12 @@ private IWritableMemChunkGroup createAlignedMemChunkGroupIfNotExistAndGet(
 
   @Override
   public void insert(InsertRowPlan insertRowPlan) {
+    // if this insert plan isn't from storage engine, we should set a temp device id for it
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java:139]:
```suggestion
    // if this insert plan isn't from storage engine(currently may be from wal), we should set a temp device id for it
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/ILastFlushTimeManager.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.engine.storagegroup;
+
+import java.util.Map;
+
+/**
+ * This interface manages last time and flush time for sequence and unsequence determination This
+ * class
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/ILastFlushTimeManager.java:25]: what is "determination this class" ?

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/LastFlushTimeManager.java [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * This class manages last time and flush time for sequence and unsequence determination This class
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/LastFlushTimeManager.java:30]:
```suggestion
 * This class manages last time and flush time for sequence and unsequence
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 11]
```diff
@@ -1360,4 +1366,19 @@ public void submitAFlushTask() {
...
   }
+
+  private IDeviceID getDeviceID(String deviceId) {
+    try {
+      return DeviceIDFactory.getInstance().getDeviceID(new PartialPath(deviceId));
+    } catch (IllegalPathException e) {
+      logger.error("device id is illegal");
+    }
+
+    return null;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:1377]: how about throwing an exception

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 12]
```diff
@@ -515,7 +517,18 @@ public void dropContinuousQuery(DropContinuousQueryPlan plan) throws IOException
   // including create and delete
 
   public void createTimeseries(CreateTimeSeriesPlan plan) throws MetadataException {
-    createTimeseries(plan, -1);
+    createTimeseriesEntry(plan, -1);
+  }
+
+  public void createTimeseriesEntry(CreateTimeSeriesPlan plan, long offset)
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:523]:
```suggestion
  public void createTimeseries(CreateTimeSeriesPlan plan, long offset)
```

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 13]
```diff
@@ -615,11 +628,22 @@ public void createAlignedTimeSeries(
...
       throws MetadataException {
-    createAlignedTimeSeries(
+    createAlignedTimeSeriesEntry(
         new CreateAlignedTimeSeriesPlan(
             prefixPath, measurements, dataTypes, encodings, compressors, null));
   }
 
+  public void createAlignedTimeSeriesEntry(CreateAlignedTimeSeriesPlan plan)
+      throws MetadataException {
+    createAlignedTimeSeries(plan);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:638]: rename this method

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 14]
```diff
@@ -615,11 +628,22 @@ public void createAlignedTimeSeries(
       List<TSEncoding> encodings,
       List<CompressionType> compressors)
       throws MetadataException {
-    createAlignedTimeSeries(
+    createAlignedTimeSeriesEntry(
         new CreateAlignedTimeSeriesPlan(
             prefixPath, measurements, dataTypes, encodings, compressors, null));
   }
 
+  public void createAlignedTimeSeriesEntry(CreateAlignedTimeSeriesPlan plan)
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:636]:
```suggestion
  public void createAlignedTimeSeries(CreateAlignedTimeSeriesPlan plan)
```

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+  private static final Logger logger = LoggerFactory.getLogger(AppendOnlyDiskSchemaManager.class);
+
+  public AppendOnlyDiskSchemaManager(File dir) {
+
+    try {
+      initFile(dir);
+      outputStream = new FileOutputStream(dataFile);
+    } catch (IOException e) {
+      logger.error(e.getMessage());
+    }
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:58]: throw ?

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+    return true;
+  }
+
+  @Override
+  public long serialize(DiskSchemaEntry schemaEntry) {
+    try {
+      schemaEntry.serialize(outputStream);
+    } catch (IOException e) {
+      logger.error("failed to serialize schema entry: " + schemaEntry);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:118]: throw

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+    return res;
+  }
+
+  @Override
+  public void close() {
+    try {
+      outputStream.close();
+    } catch (IOException e) {
+      logger.error("close schema file failed");
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:152]: throw

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTable.java [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+      }
+
+      // normal path
+      TimeseriesID timeseriesID = new TimeseriesID(fullPath);
+      return new MeasurementPath(
+          timeseriesID.getDeviceID().toStringID(),
+          timeseriesID.getMeasurement(),
+          fullPath.getMeasurementSchema());
+    } catch (MetadataException e) {
+      logger.error("Error when translate query path: " + fullPath);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTable.java:170]: throw

File: server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java [hunk 19]
```diff
@@ -299,7 +317,42 @@ public void write(TimeValuePair pair) {
...
+    private PartialPath fullPath;
+
+    IDTableLastCacheAccessor(PartialPath seriesPath) {
+      fullPath = seriesPath;
+    }
+
+    @Override
+    public TimeValuePair read() {
+      try {
+        IDTable table = IDTableManager.getInstance().getIDTable(fullPath.getDevicePath());
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java:331]: this is a global lock, could we use the StorageGroupProcessor to get the IDTable?

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTable.java [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+      return fullPath;
+    }
+
+    try {
+      // handle aligned path
+      if (fullPath instanceof AlignedPath) {
+        AlignedPath cur = (AlignedPath) fullPath;
+
+        return new AlignedPath(
+            DeviceIDFactory.getInstance().getDeviceID(cur).toStringID(),
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTable.java:158]: could we return the fullPath directly, just replace the path field.

Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTable.java:158]: I think PartialPath is an immutable class and we can't replace filed of it. There is no method to change "nodes" filed of a partial path. I think immutable is reasonable.

File: server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java [hunk 21]
```diff
@@ -1034,6 +1056,21 @@ public void mergeUnLock(List<StorageGroupProcessor> list) {
...
 
+  protected void getSeriesSchemas(InsertPlan insertPlan, StorageGroupProcessor processor)
+      throws StorageEngineException, MetadataException {
+    try {
+      if (config.isEnableIDTable()) {
+        processor.getIdTable().getSeriesSchemas(insertPlan);
+      } else {
+        IoTDB.metaManager.getSeriesSchemasAndReadLockDevice(insertPlan);
+        insertPlan.setDeviceID(DeviceIDFactory.getInstance().getDeviceID(insertPlan.getDeviceId()));
+        insertPlan.setDevicePath(insertPlan.getDeviceId());
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/StorageEngine.java:1067]: If disable idtable. We also use device id as key in memtable, so we have to set deviceID. And we need to fill device path for wal. So we just copy the reference of device path.

File: server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java (hunk 19 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java:331]: Sure, fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/entry/SHA256DeviceID.java [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+
+  /** number of message digest, for improve parallelism */
+  private static final int MD_NUM = 256;
+
+  /** logger */
+  private static Logger logger = LoggerFactory.getLogger(IDTable.class);
+
+  static {
+    try {
+      md = new MessageDigest[256];
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/entry/SHA256DeviceID.java:50]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/entry/SHA256DeviceID.java [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+      for (int i = 0; i < 256; i++) {
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/entry/SHA256DeviceID.java:51]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java (hunk 8 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/memtable/AbstractMemTable.java:141]: Wal is also from storage engine. Currently is from test.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/ILastFlushTimeManager.java (hunk 9 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/ILastFlushTimeManager.java:25]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/LastFlushTimeManager.java (hunk 10 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/LastFlushTimeManager.java:30]: fixed~

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java (hunk 11 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:1377]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java (hunk 15 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:58]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java (hunk 16 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:118]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java (hunk 17 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/AppendOnlyDiskSchemaManager.java:152]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTable.java (hunk 18 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/idtable/IDTable.java:170]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java (hunk 13 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:638]: renamed to createAlignedTimeSeriesAndCheckIDTable;

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java (hunk 14 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:636]: fixed~

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java (hunk 12 above)
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:523]: fixed~

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/query/executor/RawDataQueryExecutor.java [hunk 24]
```diff
@@ -123,6 +128,8 @@ public final QueryDataSet executeNonAlign(QueryContext context)
                 queryPlan.isAscending());
         readersOfSelectedSeries.add(reader);
       }
+    } catch (Exception e) {
+      logger.error("Meet error when init series reader " + e);
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/query/executor/RawDataQueryExecutor.java:132]: fixed~

**Review by JackieTien97 — APPROVED** [review]

**Review by ericpai — APPROVED** [review]

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**